File: oxidize-golang/core/gpucluster/gpucluster.go
// Package gpucluster models the Oxidize GPU cluster specification
// (docs/gpu_cluster_spec.md) as code: typed GPU tier profiles, Kubernetes/Helm
// manifest generation, and runtime GPU detection via nvidia-smi.
//
// It is a feature-parity port of the Rust oxidize_core::gpu_cluster module.
// YAML is emitted via string building to keep the package dependency-free.
package gpucluster

import (
	"os/exec"
	"strconv"
	"strings"
)

// Family is one of the three GPU tiers the cluster targets.
type Family int

const (
	B200 Family = iota
	A100
	RTXPro6000
)

// AllFamilies returns the families in spec order.
func AllFamilies() []Family { return []Family{B200, A100, RTXPro6000} }

// Slug returns the oxidize.io/gpu-family label value.
func (f Family) Slug() string {
	switch f {
	case B200:
		return "b200"
	case A100:
		return "a100"
	case RTXPro6000:
		return "rtx-pro-6000"
	default:
		return "unknown"
	}
}

// FamilyFromSlug parses a family from its label value, case-insensitively.
// The second return value is false when the slug is unrecognized.
func FamilyFromSlug(s string) (Family, bool) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "b200":
		return B200, true
	case "a100":
		return A100, true
	case "rtx-pro-6000", "rtx-pro6000", "rtxpro6000":
		return RTXPro6000, true
	default:
		return 0, false
	}
}

// Profile is the static hardware/scheduling profile for a GPU tier.
type Profile struct {
	Family            Family
	Product           string // NVML product name
	Generation        string // architecture shorthand
	MemoryMiB         uint32
	TDPWatts          uint32
	NVLink            bool
	MIGCapable        bool
	TimeSliceReplicas uint32
	NetworkClass      string
	WorkloadType      string
}

// ProfileFor returns the canonical profile for a family.
func ProfileFor(f Family) Profile {
	switch f {
	case B200:
		return Profile{B200, "NVIDIA-B200", "blackwell", 196608, 1000, true, false, 1, "infiniband", "training"}
	case A100:
		return Profile{A100, "NVIDIA-A100-SXM4-80GB", "ampere", 81920, 400, true, true, 2, "infiniband", "mixed"}
	case RTXPro6000:
		return Profile{RTXPro6000, "NVIDIA-RTX-Pro-6000", "ada", 98304, 300, false, false, 8, "ethernet", "workstation"}
	default:
		return Profile{}
	}
}

// AllProfiles returns the profiles for every family.
func AllProfiles() []Profile {
	out := make([]Profile, 0, 3)
	for _, f := range AllFamilies() {
		out = append(out, ProfileFor(f))
	}
	return out
}

// NodePoolSpec sizes a node pool of a given GPU family.
type NodePoolSpec struct {
	Family     Family
	NodeCount  uint32
	GPUPerNode uint32
}

func poolName(f Family) string {
	switch f {
	case B200:
		return "b200-training"
	case A100:
		return "a100-mixed"
	case RTXPro6000:
		return "rtx-pro6000"
	default:
		return "unknown"
	}
}

// NodePoolYAML renders the node-pool stanza for a single pool (spec §3.1).
func NodePoolYAML(spec NodePoolSpec) string {
	p := ProfileFor(spec.Family)
	var b strings.Builder
	b.WriteString("  " + poolName(spec.Family) + ":\n")
	b.WriteString("    count: " + strconv.FormatUint(uint64(spec.NodeCount), 10) + "\n")
	b.WriteString("    gpuPerNode: " + strconv.FormatUint(uint64(spec.GPUPerNode), 10) + "\n")
	b.WriteString("    labels:\n")
	b.WriteString("      oxidize.io/gpu-family: " + p.Family.Slug() + "\n")
	b.WriteString("      oxidize.io/gpu-arch: " + p.Generation + "\n")
	b.WriteString("      oxidize.io/workload-type: " + p.WorkloadType + "\n")
	b.WriteString("      oxidize.io/network-class: " + p.NetworkClass + "\n")
	b.WriteString("    taints:\n")
	b.WriteString("      - key: oxidize.io/gpu\n")
	b.WriteString("        value: " + p.Family.Slug() + "\n")
	b.WriteString("        effect: NoSchedule\n")
	return b.String()
}

// NodePoolsYAML renders the full nodePools document.
func NodePoolsYAML(specs []NodePoolSpec) string {
	var b strings.Builder
	b.WriteString("nodePools:\n")
	for _, s := range specs {
		b.WriteString(NodePoolYAML(s))
	}
	return b.String()
}

// Label is a key/value node label.
type Label struct{ Key, Value string }

// NodeLabels renders the GFD/scheduling labels a node must carry (spec §3.2).
func NodeLabels(f Family, gpuCount uint32) []Label {
	p := ProfileFor(f)
	migCapable := "false"
	if p.MIGCapable {
		migCapable = "true"
	}
	return []Label{
		{"nvidia.com/gpu.present", "true"},
		{"nvidia.com/gpu.product", p.Product},
		{"nvidia.com/gpu.count", strconv.FormatUint(uint64(gpuCount), 10)},
		{"nvidia.com/gpu.memory", strconv.FormatUint(uint64(p.MemoryMiB), 10)},
		{"nvidia.com/mig.capable", migCapable},
		{"oxidize.io/gpu-family", p.Family.Slug()},
		{"oxidize.io/gpu-generation", p.Generation},
		{"oxidize.io/network-class", p.NetworkClass},
	}
}

// DevicePluginConfigYAML renders the device-plugin time-slicing ConfigMap (§4.3).
func DevicePluginConfigYAML(families []Family) string {
	var b strings.Builder
	b.WriteString("apiVersion: v1\n")
	b.WriteString("kind: ConfigMap\n")
	b.WriteString("metadata:\n")
	b.WriteString("  name: nvidia-device-plugin-config\n")
	b.WriteString("  namespace: kube-system\n")
	b.WriteString("data:\n")
	b.WriteString("  config.yaml: |\n")
	b.WriteString("    version: v1\n")
	b.WriteString("    sharing:\n")
	b.WriteString("      timeSlicing:\n")
	b.WriteString("        renameByDefault: false\n")
	b.WriteString("        failRequestsGreaterThanOne: true\n")
	b.WriteString("        resources:\n")
	b.WriteString("          - name: nvidia.com/gpu\n")
	b.WriteString("            replicas: 1\n")

	var overrides []Family
	for _, f := range families {
		if ProfileFor(f).TimeSliceReplicas > 1 {
			overrides = append(overrides, f)
		}
	}
	if len(overrides) > 0 {
		b.WriteString("    nodes:\n")
		for _, f := range overrides {
			p := ProfileFor(f)
			b.WriteString("      - match:\n")
			b.WriteString("          - key: oxidize.io/gpu-family\n")
			b.WriteString("            operator: In\n")
			b.WriteString("            values:\n")
			b.WriteString("              - " + p.Family.Slug() + "\n")
			b.WriteString("        sharing:\n")
			b.WriteString("          timeSlicing:\n")
			b.WriteString("            renameByDefault: true\n")
			b.WriteString("            resources:\n")
			b.WriteString("              - name: nvidia.com/gpu\n")
			b.WriteString("                replicas: " + strconv.FormatUint(uint64(p.TimeSliceReplicas), 10) + "\n")
		}
	}
	return b.String()
}

// MIGProfile is a recommended MIG geometry (spec §4.4).
type MIGProfile struct {
	Name       string
	MemoryGB   uint32
	ComputeSMs uint32
	BestFor    string
}

// MIGProfiles returns the recommended A100 MIG geometries.
func MIGProfiles() []MIGProfile {
	return []MIGProfile{
		{"1g.10gb", 10, 14, "light inference, micro-services"},
		{"2g.20gb", 20, 28, "medium inference, small training"},
		{"3g.40gb", 40, 42, "large model inference, fine-tuning"},
		{"4g.40gb", 40, 56, "heavy inference, data processing"},
		{"7g.80gb", 80, 108, "large training (disable MIG)"},
	}
}

// MIGConfigYAML renders the MIG strategy ConfigMap, or ("", false) when the
// family is not MIG-capable.
func MIGConfigYAML(f Family) (string, bool) {
	p := ProfileFor(f)
	if !p.MIGCapable {
		return "", false
	}
	var b strings.Builder
	b.WriteString("apiVersion: v1\n")
	b.WriteString("kind: ConfigMap\n")
	b.WriteString("metadata:\n")
	b.WriteString("  name: nvidia-mig-config\n")
	b.WriteString("  namespace: kube-system\n")
	b.WriteString("data:\n")
	b.WriteString("  config.yaml: |\n")
	b.WriteString("    version: v1\n")
	b.WriteString("    flags:\n")
	b.WriteString("      migStrategy: mixed\n")
	b.WriteString("    nodes:\n")
	b.WriteString("      - match:\n")
	b.WriteString("          - key: nvidia.com/gpu.product\n")
	b.WriteString("            operator: In\n")
	b.WriteString("            values:\n")
	b.WriteString("              - " + p.Product + "\n")
	b.WriteString("        mig:\n")
	b.WriteString("          strategy: mixed\n")
	return b.String(), true
}

// GPUTolerationsYAML renders the standard pod tolerations (spec §3.3).
func GPUTolerationsYAML() string {
	return "tolerations:\n" +
		"  - key: \"oxidize.io/gpu\"\n" +
		"    operator: \"Exists\"\n" +
		"    effect: \"NoSchedule\"\n" +
		"  - key: \"nvidia.com/gpu\"\n" +
		"    operator: \"Exists\"\n" +
		"    effect: \"NoSchedule\"\n"
}

// HelmValuesYAML renders the GPU-Operator Helm values for a family (spec §5.2).
func HelmValuesYAML(f Family) string {
	p := ProfileFor(f)
	driver := "535.161.08"
	open := "false"
	if p.Generation == "blackwell" {
		driver = "550.54.15"
		open = "true"
	}
	mig := "false"
	if p.MIGCapable {
		mig = "true"
	}
	var b strings.Builder
	b.WriteString("# GPU-Operator values for the " + p.Family.Slug() + " (" + p.Generation + ") node pool\n")
	b.WriteString("driver:\n")
	b.WriteString("  enabled: true\n")
	b.WriteString("  version: \"" + driver + "\"\n")
	b.WriteString("  useOpenKernelModules: " + open + "\n")
	b.WriteString("toolkit:\n")
	b.WriteString("  enabled: true\n")
	b.WriteString("devicePlugin:\n")
	b.WriteString("  enabled: true\n")
	b.WriteString("  config:\n")
	b.WriteString("    name: nvidia-device-plugin-config\n")
	b.WriteString("dcgmExporter:\n")
	b.WriteString("  enabled: true\n")
	b.WriteString("  serviceMonitor:\n")
	b.WriteString("    enabled: true\n")
	b.WriteString("migManager:\n")
	b.WriteString("  enabled: " + mig + "\n")
	return b.String()
}

// PrometheusRulesYAML renders the DCGM PrometheusRule for GPU alerts (§8.1).
func PrometheusRulesYAML() string {
	return "apiVersion: monitoring.coreos.com/v1\n" +
		"kind: PrometheusRule\n" +
		"metadata:\n" +
		"  name: oxidize-gpu-alerts\n" +
		"spec:\n" +
		"  groups:\n" +
		"    - name: gpu-health\n" +
		"      rules:\n" +
		"        - alert: GPUHighTemperature\n" +
		"          expr: dcgm_gpu_temp > 85\n" +
		"          for: 5m\n" +
		"          labels:\n" +
		"            severity: critical\n" +
		"        - alert: GPUMemoryNearExhaustion\n" +
		"          expr: dcgm_fb_used / (dcgm_fb_free + dcgm_fb_used) > 0.95\n" +
		"          for: 10m\n" +
		"          labels:\n" +
		"            severity: warning\n" +
		"        - alert: NVLinkError\n" +
		"          expr: dcgm_nvlink_replay_error_count_total > 0\n" +
		"          for: 1m\n" +
		"          labels:\n" +
		"            severity: critical\n"
}

// DetectedGPU is a physical GPU discovered via nvidia-smi.
type DetectedGPU struct {
	Index          uint32
	Name           string
	MemoryTotalMiB uint32
	MIGEnabled     bool
	Family         Family // valid only when FamilyKnown is true
	FamilyKnown    bool
}

// ClassifyProduct classifies an NVML/nvidia-smi product name into a Family.
// The second return value is false when the name is unrecognized.
func ClassifyProduct(name string) (Family, bool) {
	n := strings.ToLower(name)
	switch {
	case strings.Contains(n, "b200"):
		return B200, true
	case strings.Contains(n, "a100"):
		return A100, true
	case (strings.Contains(n, "rtx") || strings.Contains(n, "pro")) && strings.Contains(n, "6000"):
		return RTXPro6000, true
	default:
		return 0, false
	}
}

// ParseNvidiaSMICSV parses the CSV output of
// nvidia-smi --query-gpu=index,name,memory.total,mig.mode.current
// --format=csv,noheader,nounits. Unparseable lines are skipped.
func ParseNvidiaSMICSV(output string) []DetectedGPU {
	var gpus []DetectedGPU
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		fields := strings.Split(line, ",")
		for i := range fields {
			fields[i] = strings.TrimSpace(fields[i])
		}
		if len(fields) < 3 {
			continue
		}
		index, err := strconv.ParseUint(fields[0], 10, 32)
		if err != nil {
			continue
		}
		mem, err := strconv.ParseUint(fields[2], 10, 32)
		if err != nil {
			continue
		}
		migEnabled := len(fields) > 3 && strings.EqualFold(fields[3], "enabled")
		fam, known := ClassifyProduct(fields[1])
		gpus = append(gpus, DetectedGPU{
			Index:          uint32(index),
			Name:           fields[1],
			MemoryTotalMiB: uint32(mem),
			MIGEnabled:     migEnabled,
			Family:         fam,
			FamilyKnown:    known,
		})
	}
	return gpus
}

// DetectGPUs probes the local node for NVIDIA GPUs via nvidia-smi. It returns
// an empty slice when nvidia-smi is unavailable or fails.
func DetectGPUs() []DetectedGPU {
	out, err := exec.Command("nvidia-smi",
		"--query-gpu=index,name,memory.total,mig.mode.current",
		"--format=csv,noheader,nounits").Output()
	if err != nil {
		return []DetectedGPU{}
	}
	return ParseNvidiaSMICSV(string(out))
}

// FamilyCount pairs a family with a count.
type FamilyCount struct {
	Family Family
	Count  uint32
}

// Summarize counts detected GPUs by family, in spec order.
func Summarize(gpus []DetectedGPU) []FamilyCount {
	counts := map[Family]uint32{}
	for _, g := range gpus {
		if g.FamilyKnown {
			counts[g.Family]++
		}
	}
	var out []FamilyCount
	for _, f := range AllFamilies() {
		if n := counts[f]; n > 0 {
			out = append(out, FamilyCount{f, n})
		}
	}
	// out is already in AllFamilies spec order; no extra sort needed.
	return out
}

File: oxidize-c/src/cluster/gpu_cluster.c
/*
 * gpu_cluster.c — GPU family profiles and Kubernetes manifest generation.
 *
 * Port from oxidize-core/src/cluster/gpu_cluster.rs.
 */
#define _POSIX_C_SOURCE 200809L

#include "oxidize/gpu_cluster.h"

#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h> /* strcasecmp */

/* ------------------------------------------------------------------ */
/* Static GPU family profiles.                                         */
/* ------------------------------------------------------------------ */

static const OcGpuProfile OC_PROFILES[] = {
    [OC_GPU_FAMILY_B200] = {
        .family = OC_GPU_FAMILY_B200,
        .product = "NVIDIA-B200",
        .generation = "blackwell",
        .memory_mib = 196608u,
        .tdp_watts = 1000u,
        .nvlink = true,
        .mig_capable = false,
        .time_slice_replicas = 1u,
        .network_class = "infiniband",
        .workload_type = "training",
    },
    [OC_GPU_FAMILY_A100] = {
        .family = OC_GPU_FAMILY_A100,
        .product = "NVIDIA-A100-SXM4-80GB",
        .generation = "ampere",
        .memory_mib = 81920u,
        .tdp_watts = 400u,
        .nvlink = true,
        .mig_capable = true,
        .time_slice_replicas = 2u,
        .network_class = "infiniband",
        .workload_type = "inference",
    },
    [OC_GPU_FAMILY_RTX_PRO_6000] = {
        .family = OC_GPU_FAMILY_RTX_PRO_6000,
        .product = "NVIDIA-RTX-PRO-6000",
        .generation = "ada",
        .memory_mib = 49152u,
        .tdp_watts = 300u,
        .nvlink = false,
        .mig_capable = false,
        .time_slice_replicas = 4u,
        .network_class = "ethernet",
        .workload_type = "edge",
    },
    [OC_GPU_FAMILY_H100] = {
        .family = OC_GPU_FAMILY_H100,
        .product = "NVIDIA-H100-SXM5-80GB",
        .generation = "hopper",
        .memory_mib = 81920u,
        .tdp_watts = 700u,
        .nvlink = true,
        .mig_capable = false,
        .time_slice_replicas = 1u,
        .network_class = "infiniband",
        .workload_type = "throughput-inference",
    },
};

static const char *const OC_FAMILY_SLUGS[] = {
    [OC_GPU_FAMILY_B200] = "b200",
    [OC_GPU_FAMILY_A100] = "a100",
    [OC_GPU_FAMILY_RTX_PRO_6000] = "rtx-pro-6000",
    [OC_GPU_FAMILY_H100] = "h100",
};

static const char *const OC_FAMILY_NAMES[] = {
    [OC_GPU_FAMILY_B200] = "B200",
    [OC_GPU_FAMILY_A100] = "A100",
    [OC_GPU_FAMILY_RTX_PRO_6000] = "RTX Pro 6000",
    [OC_GPU_FAMILY_H100] = "H100",
};

/* Higher rank = more capable. B200 is the flagship, then H100, A100,
 * RTX Pro 6000. */
static const uint8_t OC_FAMILY_RANKS[] = {
    [OC_GPU_FAMILY_B200] = 4u,
    [OC_GPU_FAMILY_A100] = 2u,
    [OC_GPU_FAMILY_RTX_PRO_6000] = 1u,
    [OC_GPU_FAMILY_H100] = 3u,
};

#define OC_N_FAMILIES \
    (sizeof(OC_PROFILES) / sizeof(OC_PROFILES[0]))

/* ------------------------------------------------------------------ */
/* Helpers.                                                            */
/* ------------------------------------------------------------------ */

static bool family_valid(OcGpuFamily f)
{
    return (size_t)f < OC_N_FAMILIES;
}

/* ------------------------------------------------------------------ */
/* Public API.                                                         */
/* ------------------------------------------------------------------ */

const OcGpuProfile *oc_gpu_profile(OcGpuFamily family)
{
    if (!family_valid(family)) return NULL;
    return &OC_PROFILES[family];
}

const char *oc_gpu_family_slug(OcGpuFamily family)
{
    if (!family_valid(family)) return "unknown";
    return OC_FAMILY_SLUGS[family];
}

const char *oc_gpu_family_name(OcGpuFamily family)
{
    if (!family_valid(family)) return "unknown";
    return OC_FAMILY_NAMES[family];
}

OcGpuFamily oc_gpu_family_from_slug(const char *slug)
{
    if (!slug) return OC_GPU_FAMILY__COUNT;
    for (size_t i = 0; i < OC_N_FAMILIES; i++) {
        if (strcasecmp(slug, OC_FAMILY_SLUGS[i]) == 0)
            return (OcGpuFamily)i;
    }
    return OC_GPU_FAMILY__COUNT;
}

uint8_t oc_gpu_family_rank(OcGpuFamily family)
{
    if (!family_valid(family)) return 0u;
    return OC_FAMILY_RANKS[family];
}

size_t oc_gpu_n_families(void)
{
    return OC_N_FAMILIES;
}

OcGpuFamily oc_gpu_family_by_index(size_t idx)
{
    if (idx >= OC_N_FAMILIES) return OC_GPU_FAMILY__COUNT;
    return (OcGpuFamily)idx;
}

const char *oc_gpu_cluster_label(const char *key)
{
    if (!key) return NULL;
    /* The label function returns the value for the B200 family by default,
     * mirroring the Rust "default tier" behavior. Callers that need per-family
     * label values can read the profile directly. */
    const OcGpuProfile *p = &OC_PROFILES[OC_GPU_FAMILY_B200];
    static char mem_buf[32];
    static char count_buf[32];

    if (strcmp(key, "nvidia.com/gpu.product") == 0) return p->product;
    if (strcmp(key, "nvidia.com/gpu.memory") == 0) {
        snprintf(mem_buf, sizeof(mem_buf), "%u", p->memory_mib);
        return mem_buf;
    }
    if (strcmp(key, "nvidia.com/gpu.count") == 0) {
        snprintf(count_buf, sizeof(count_buf), "%u", p->time_slice_replicas);
        return count_buf;
    }
    if (strcmp(key, "network") == 0) return p->network_class;
    if (strcmp(key, "workload") == 0) return p->workload_type;
    return NULL;
}

/* ------------------------------------------------------------------ */
/* Manifest generation.                                                */
/* ------------------------------------------------------------------ */

OcError oc_gpu_cluster_node_pool_yaml(OcGpuFamily family, uint32_t replicas,
                                      char *out, size_t out_len)
{
    if (!family_valid(family)) return OC_ERR_INVALID_ARG;
    if (!out || out_len == 0) return OC_ERR_INVALID_ARG;

    const OcGpuProfile *p = &OC_PROFILES[family];
    int n = snprintf(out, out_len,
        "apiVersion: apps/v1\n"
        "kind: NodePool\n"
        "metadata:\n"
        "  name: gpu-%s-pool\n"
        "  labels:\n"
        "    nvidia.com/gpu.product: \"%s\"\n"
        "    nvidia.com/gpu.memory: \"%u\"\n"
        "    oxidize.io/gpu.family: \"%s\"\n"
        "    oxidize.io/gpu.generation: \"%s\"\n"
        "    oxidize.io/network.class: \"%s\"\n"
        "    oxidize.io/workload.type: \"%s\"\n"
        "spec:\n"
        "  replicas: %u\n"
        "  nodeSelector:\n"
        "    nvidia.com/gpu.present: \"true\"\n"
        "  tolerations:\n"
        "    - key: nvidia.com/gpu\n"
        "      operator: Exists\n"
        "  config:\n"
        "    gpu:\n"
        "      memoryMiB: %u\n"
        "      tdpWatts: %u\n"
        "      nvlink: %s\n"
        "      mig: %s\n"
        "      timeSliceReplicas: %u\n",
        OC_FAMILY_SLUGS[family],
        p->product,
        p->memory_mib,
        OC_FAMILY_SLUGS[family],
        p->generation,
        p->network_class,
        p->workload_type,
        replicas,
        p->memory_mib,
        p->tdp_watts,
        p->nvlink ? "true" : "false",
        p->mig_capable ? "true" : "false",
        p->time_slice_replicas);

    if (n < 0) return OC_ERR_INTERNAL;
    if ((size_t)n >= out_len) return OC_ERR_INTERNAL;
    return OC_OK;
}

OcError oc_gpu_cluster_device_plugin_yaml(OcGpuFamily family, char *out, size_t out_len)
{
    if (!family_valid(family)) return OC_ERR_INVALID_ARG;
    if (!out || out_len == 0) return OC_ERR_INVALID_ARG;

    const OcGpuProfile *p = &OC_PROFILES[family];
    int n = snprintf(out, out_len,
        "apiVersion: apps/v1\n"
        "kind: DaemonSet\n"
        "metadata:\n"
        "  name: nvidia-device-plugin-%s\n"
        "  namespace: kube-system\n"
        "  labels:\n"
        "    oxidize.io/gpu.family: \"%s\"\n"
        "spec:\n"
        "  selector:\n"
        "    matchLabels:\n"
        "      name: nvidia-device-plugin-%s\n"
        "  template:\n"
        "    metadata:\n"
        "      labels:\n"
        "        name: nvidia-device-plugin-%s\n"
        "    spec:\n"
        "      nodeSelector:\n"
        "        nvidia.com/gpu.product: \"%s\"\n"
        "      tolerations:\n"
        "        - key: nvidia.com/gpu\n"
        "          operator: Exists\n"
        "      containers:\n"
        "        - name: nvidia-device-plugin\n"
        "          image: nvcr.io/nvidia/k8s-device-plugin:v0.14.5\n"
        "          env:\n"
        "            - name: NVIDIA_GPU_MEMORY_MIB\n"
        "              value: \"%u\"\n"
        "            - name: NVIDIA_TIME_SLICE_REPLICAS\n"
        "              value: \"%u\"\n"
        "            - name: NVIDIA_MIG_ENABLED\n"
        "              value: \"%s\"\n"
        "          securityContext:\n"
        "            privileged: true\n",
        OC_FAMILY_SLUGS[family],
        OC_FAMILY_SLUGS[family],
        OC_FAMILY_SLUGS[family],
        OC_FAMILY_SLUGS[family],
        p->product,
        p->memory_mib,
        p->time_slice_replicas,
        p->mig_capable ? "true" : "false");

    if (n < 0) return OC_ERR_INTERNAL;
    if ((size_t)n >= out_len) return OC_ERR_INTERNAL;
    return OC_OK;
}

/* ------------------------------------------------------------------ */
/* Runtime detection (pure classifier + nvidia-smi probe).           */
/* ------------------------------------------------------------------ */

static void ascii_lower_copy(char *dst, size_t cap, const char *src)
{
    size_t i = 0;
    if (cap == 0) return;
    if (src == NULL) {
        dst[0] = '\0';
        return;
    }
    for (; src[i] != '\0' && i + 1 < cap; i++) {
        char c = src[i];
        if (c >= 'A' && c <= 'Z') c = (char)(c - 'A' + 'a');
        dst[i] = c;
    }
    dst[i] = '\0';
}

OcGpuFamily oc_gpu_family_from_nvidia_name(const char *name)
{
    char n[256];
    if (name == NULL || name[0] == '\0') return OC_GPU_FAMILY__COUNT;
    ascii_lower_copy(n, sizeof(n), name);
    if (strstr(n, "b200") != NULL) return OC_GPU_FAMILY_B200;
    if (strstr(n, "h100") != NULL) return OC_GPU_FAMILY_H100;
    if (strstr(n, "a100") != NULL) return OC_GPU_FAMILY_A100;
    if (strstr(n, "rtx") != NULL && strstr(n, "pro") != NULL &&
        strstr(n, "6000") != NULL)
        return OC_GPU_FAMILY_RTX_PRO_6000;
    return OC_GPU_FAMILY__COUNT;
}

static void trim_copy(char *dst, size_t cap, const char *src, size_t len)
{
    while (len > 0 && (*src == ' ' || *src == '\t')) {
        src++;
        len--;
    }
    while (len > 0 && (src[len - 1] == ' ' || src[len - 1] == '\t' ||
                       src[len - 1] == '\r' || src[len - 1] == '\n')) {
        len--;
    }
    if (len >= cap) len = cap - 1;
    if (len > 0) memcpy(dst, src, len);
    dst[len] = '\0';
}

static bool parse_csv_u32(const char *s, uint32_t *out)
{
    char *end = NULL;
    unsigned long v;

    if (s == NULL || s[0] == '\0' || s[0] == '+' || s[0] == '-')
        return false;
    errno = 0;
    v = strtoul(s, &end, 10);
    if (end == s || *end != '\0') return false;
    if (errno == ERANGE || v > (unsigned long)UINT32_MAX) return false;
    *out = (uint32_t)v;
    return true;
}

OcError oc_gpu_parse_nvidia_smi_csv(const char *output, OcGpuDevice *out,
                                      size_t cap, size_t *out_n)
{
    if (out_n == NULL) return OC_ERR_INVALID_ARG;
    *out_n = 0;
    if (output == NULL) return OC_OK;
    if (cap > 0 && out == NULL) return OC_ERR_INVALID_ARG;

    const char *line = output;
    while (*line != '\0' && *out_n < cap) {
        const char *eol = strchr(line, '\n');
        size_t linelen = eol ? (size_t)(eol - line) : strlen(line);
        size_t t = 0;
        while (t < linelen && (line[t] == ' ' || line[t] == '\t' ||
                                line[t] == '\r'))
            t++;
        if (t == linelen) {
            line = eol ? eol + 1 : line + linelen;
            continue;
        }

        const char *fields[4];
        size_t flen[4];
        size_t nfields = 0;
        const char *p = line;
        const char *end = line + linelen;
        while (nfields < 4) {
            const char *comma = p;
            while (comma < end && *comma != ',') comma++;
            fields[nfields] = p;
            flen[nfields] = (size_t)(comma - p);
            nfields++;
            if (comma >= end) break;
            p = comma + 1;
        }
        if (nfields < 3) {
            line = eol ? eol + 1 : line + linelen;
            continue;
        }

        char idxbuf[32], namebuf[128], membuf[32], migbuf[32];
        trim_copy(idxbuf, sizeof(idxbuf), fields[0], flen[0]);
        trim_copy(namebuf, sizeof(namebuf), fields[1], flen[1]);
        trim_copy(membuf, sizeof(membuf), fields[2], flen[2]);
        if (nfields >= 4)
            trim_copy(migbuf, sizeof(migbuf), fields[3], flen[3]);
        else
            migbuf[0] = '\0';

        uint32_t index = 0, mem = 0;
        if (!parse_csv_u32(idxbuf, &index) || !parse_csv_u32(membuf, &mem)) {
            line = eol ? eol + 1 : line + linelen;
            continue;
        }

        OcGpuDevice *d = &out[*out_n];
        memset(d, 0, sizeof(*d));
        d->index = index;
        memcpy(d->name, namebuf, strlen(namebuf) + 1);
        d->memory_total_mib = mem;
        d->mig_enabled = (strcasecmp(migbuf, "enabled") == 0);
        d->family = oc_gpu_family_from_nvidia_name(d->name);
        (*out_n)++;

        line = eol ? eol + 1 : line + linelen;
    }
    return OC_OK;
}

OcError oc_gpu_detect(OcGpuDevice *out, size_t cap, size_t *out_n)
{
    if (out_n == NULL) return OC_ERR_INVALID_ARG;
    *out_n = 0;
    if (cap > 0 && out == NULL) return OC_ERR_INVALID_ARG;

    FILE *fp = popen("nvidia-smi --query-gpu=index,name,memory.total,mig.mode.current "
                     "--format=csv,noheader,nounits 2>/dev/null",
                     "r");
    if (fp == NULL) return OC_OK;

    char buf[8192];
    size_t n = 0;
    while (n + 1 < sizeof(buf)) {
        if (fgets(buf + n, (int)(sizeof(buf) - n), fp) == NULL) break;
        n += strlen(buf + n);
    }
    int status = pclose(fp);
    if (status != 0 || n == 0) return OC_OK;
    if (n == sizeof(buf) - 1 && buf[n - 1] != '\n') {
        char *last_nl = strrchr(buf, '\n');
        if (last_nl != NULL)
            last_nl[1] = '\0';
        else
            buf[0] = '\0';
    }
    return oc_gpu_parse_nvidia_smi_csv(buf, out, cap, out_n);
}
